staging/src/k8s.io/apiserver/pkg/util/version/version.go

```diff
@@ -0,0 +1,177 @@
+	return m.emulationVersion.Val.Load().WithPreRelease(m.BinaryVersion().PreRelease())
+}
+
+func (m *effectiveVersions) MinCompatibilityVersion() *version.Version {
+	return m.minCompatibilityVersion.Val.Load()
+}
+
+func (m *effectiveVersions) Set(binaryVersion, emulationVersion, minCompatibilityVersion *version.Version) {
+	m.binaryVersion.Store(binaryVersion)
+	m.emulationVersion.Val.Store(emulationVersion)
+	m.minCompatibilityVersion.Val.Store(minCompatibilityVersion)
+}
+
+func (m *effectiveVersions) SetBinaryVersionForTests(binaryVersion *version.Version) func() {
+	oldBinaryVersion := m.binaryVersion.Load()
+	m.Set(binaryVersion, binaryVersion, version.MajorMinor(binaryVersion.Major(), binaryVersion.SubtractMinor(1).Minor()))
+	oldFeatureGateVersion := utilfeature.DefaultFeatureGate.EmulationVersion()
+	if err := utilfeature.DefaultMutableVersionedFeatureGate.SetEmulationVersion(binaryVersion); err != nil {
+		panic(err)
+	}
+	return func() {
+		m.Set(oldBinaryVersion, oldBinaryVersion, version.MajorMinor(oldBinaryVersion.Major(), oldBinaryVersion.SubtractMinor(1).Minor()))
+		if err := utilfeature.DefaultMutableVersionedFeatureGate.SetEmulationVersion(oldFeatureGateVersion); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (m *effectiveVersions) Validate() []error {
+	var errs []error
+	// Validate only checks the major and minor versions.
+	binaryVersion := m.binaryVersion.Load().WithPatch(0)
+	emulationVersion := m.emulationVersion.Val.Load().WithPatch(0)
+	minCompatibilityVersion := m.minCompatibilityVersion.Val.Load().WithPatch(0)
+
+	maxEmuVer := binaryVersion
+	minEmuVer := binaryVersion
+	// emulationVersion can only be 1.{binaryMinor-1}...1.{binaryMinor} for alpha if EmulationVersion feature gate is enabled.
+	// otherwise, emulationVersion can only be 1.{binaryMinor}.
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.EmulationVersion) {
+		minEmuVer = binaryVersion.SubtractMinor(1)
+	}
+	// emulationVersion concept is introduced in 1.30, it cannot be set to be less than that.
+	if minEmuVer.LessThan(version.MajorMinor(1, 30)) {
+		minEmuVer = version.MajorMinor(1, 30)
+	}
+	if emulationVersion.GreaterThan(maxEmuVer) || emulationVersion.LessThan(minEmuVer) {
+		errs = append(errs, fmt.Errorf("emulation version %s is not between [%s, %s]", emulationVersion.String(), minEmuVer.String(), maxEmuVer.String()))
+	}
+	// minCompatibilityVersion can only be 1.{binaryMinor-1} for alpha.
+	maxCompVer := binaryVersion.SubtractMinor(1)
+	minCompVer := binaryVersion.SubtractMinor(1)
+	if minCompatibilityVersion.GreaterThan(maxCompVer) || minCompatibilityVersion.LessThan(minCompVer) {
+		errs = append(errs, fmt.Errorf("minCompatibilityVersion version %s is not between [%s, %s]", minCompatibilityVersion.String(), minCompVer.String(), maxCompVer.String()))
+	}
+	return errs
+}
+
+func (m *effectiveVersions) AddFlags(fs *pflag.FlagSet) {
+	if m == nil {
+		return
+	}
+
+	fs.Var(&m.emulationVersion, "emulated-version", ""+
```

Comment:
I expected that if the user sets the emulationVersion without changing `minCompatibilityVersion`, we would assume `N-1`, is that not the case? How would user specify `minCompatibiltyVersion` or would they not?

Reply:
We are not addressing `minCompatibiltyVersion` in 1.30. The flag would be added in the next version.